### PR TITLE
feat(auth): add signInDetails to AuthUser interface

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

Adds signInDetails to the `AuthUser` interface. This change will provide information used during the sign-in flow, such as `loginId` and `authFlowType`

**DX:**
1. userpool configure to sign-in with either phone_number or email.
Screenshot 2023-10-26 at 9 36 48 AM

2. userpool configure to sign-in with alias
Screenshot 2023-10-26 at 9 38 03 AM

#### Description of how you validated changes

- unit test
- smoke test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts [hunk 1]
```diff
@@ -1,22 +1,38 @@
...
+	if (isTokensWithSignInDetails(tokens)) {
+		authUser.signInDetails = tokens.signInDetails;
+	}
+	return authUser;
 };
+
+function isTokensWithSignInDetails(
+	tokens: Record<string, unknown>
+): tokens is AuthTokens & { signInDetails: CognitoAuthSignInDetails } {
+	return !!tokens && !!(tokens as any).signInDetails;
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts:37]: Do you need to check what `signInDetails`?

File: packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts [hunk 2]
```diff
@@ -65,14 +65,22 @@ export class DefaultTokenStore implements AuthTokenStore {
...
 				accessToken,
 				idToken,
 				refreshToken,
 				deviceMetadata: (await this.getDeviceMetadata()) ?? undefined,
 				clockDrift,
 				username: await this.getLastAuthUser(),
 			};
+
+			if (signInDetails) {
+				tokens.signInDetails = JSON.parse(signInDetails);
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts:81]: Do we need to check if this throws?

File: packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts [hunk 3]
```diff
@@ -125,6 +133,12 @@ export class DefaultTokenStore implements AuthTokenStore {
 				tokens.deviceMetadata.randomPassword
 			);
 		}
+		if (!!tokens.signInDetails) {
+			await this.getKeyValueStorage().setItem(
+				authKeys.signInDetails,
+				JSON.stringify(tokens.signInDetails)
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts:139]: Do we need to check if this throws?

File: packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts (hunk 2 above)
Comment (jimblanc) [reply packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts:81]: Looks like it's wrapped in a try-catch if that's adequate?

**Review by jimblanc — DISMISSED** [review]
LGTM

File: packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts (hunk 1 above)
Comment (cshfang) [inline packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts:37]: Nit: Could be more concise
```ts
	return !!(tokens as any)?.signInDetails;
```

File: packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts [hunk 4]
```diff
@@ -1,22 +1,38 @@
...
 		username: username as string,
 		userId: sub as string,
 	};
+	if (isTokensWithSignInDetails(tokens)) {
+		authUser.signInDetails = tokens.signInDetails;
+	}
+	return authUser;
 };
+
+function isTokensWithSignInDetails(
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts:34]: Nit: Does this need to be a type predicate? Is there any reason above to narrow type? I'm just wondering if this function can be something like `getSignInDetailsFromTokens` and just return either `CognitoAuthSignInDetails | undefined` - which would help consolidate some logic above

File: packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts [hunk 5]
```diff
@@ -62,7 +63,9 @@ export class TokenOrchestrator implements AuthTokenOrchestrator {
 
 	async getTokens(
 		options?: FetchAuthSessionOptions
-	): Promise<AuthTokens | null> {
+	): Promise<
+		(AuthTokens & { signInDetails?: CognitoAuthSignInDetails }) | null
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:67]: Nit: I'm sure there was a reason for this but why can't `signInDetails` just be an optional property on `AuthTokens`?

File: packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts (hunk 2 above)
Comment (elorzafe) [reply packages/auth/src/providers/cognito/tokenProvider/TokenStore.ts:81]: if there is something that was modified I would handle that there. can you change that @israx ?

File: packages/auth/src/providers/cognito/types/outputs.ts [hunk 6]
```diff
@@ -29,7 +29,7 @@ export type FetchUserAttributesOutput = AuthUserAttributes<UserAttributeKey>;
 /**
  * Output type for Cognito getCurrentUser API.
  */
-export type GetCurrentUserOutput = AuthUser;
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/types/outputs.ts:32]: Nit: Is `AuthUser` still used?

**Review by cshfang — DISMISSED** [review]
Had some questions/suggestions but I don't _think_ they're blocking

**Review by elorzafe — APPROVED** [review]
Looks good it seems weird that `signIndetails` gets modified externally

File: packages/auth/src/providers/cognito/types/outputs.ts (hunk 6 above)
Comment (israx) [reply packages/auth/src/providers/cognito/types/outputs.ts:32]: It is used to extend the `CognitoAuthUser`

File: packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts (hunk 5 above)
Comment (israx) [reply packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:67]: So if customers are extending the TokenProvider interface, they don't have to add signInDetails

File: packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts (hunk 1 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts:37]: I think for now, this should be good, thinking if in the feature signInDetails gets extended, then we would need to account for it too

Comment (israx) [reply packages/auth/src/providers/cognito/apis/internal/getCurrentUser.ts:37]: changed the logic to include @cshfang feedback, this code is not longer valid.

**Review by jimblanc — APPROVED** [review]

**Review by elorzafe — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

**Pull request merged**